### PR TITLE
[WIP] Upgrade to vega 3, vega-lite 2, and vega-embed 3

### PR DESCRIPTION
**Pull request by gnestor**
Closes https://github.com/altair-viz/jupyterlab_vega/issues/23

### CODE REVIEW

**Comment by gnestor** [comment]
@domoritz I'm getting the following error in classic notebook using vega-embed@3.0.0-beta.5:

```
Uncaught SyntaxError: Unexpected token import  event-selector.js:1
```

Which is causing the extension to fail to load:

```
Loading extension: jupyterlab_vega/extension  utils.js:37 
TypeError: Cannot read property 'register_renderer' of undefined  require.js?v=6da8be3…:900
    at Object.<anonymous> (extension.js:31)
    at Object.load_ipython_extension (extension.js:24)
    at utils.js:39
    at Object.execCb (require.js?v=6da8be3…:1690)
    at Module.check (require.js?v=6da8be3…:865)
    at Module.<anonymous> (require.js?v=6da8be3…:1140)
    at require.js?v=6da8be3…:131
    at require.js?v=6da8be3…:1190
    at each (require.js?v=6da8be3…:56)
    at Module.emit (require.js?v=6da8be3…:1189)
```

Downgrading to beta4 resolves it, so it's probably related to https://github.com/vega/vega-embed/commit/3fbc42fef837d17b40df85ead427723440d3aa7e,

**Comment by domoritz** [comment]
Vega embed 3 requires Vega 2 and Vega-Lite 2. Let me try and see whether it works locally for me.

**Comment by gnestor** [comment]
Do you mean Vega 3? According to the `package.json`:

```
"vega": "^3.0.0-beta.25",
"vega-lite": "^2.0.0-alpha.4"
```

File: component/package.json [hunk 1]
```diff
@@ -13,10 +13,11 @@
...
-    "d3": "3.5.17",
+    "d3-selection": "^1.0.4",
     "react": "^15.3.2",
-    "vega": "^2.6.1",
-    "vega-embed": "^2.2.0",
-    "vega-lite": "^1.1.3"
+    "vega": "^3.0.0-beta.25",
+    "vega-lite": "^2.0.0-alpha.4",
+    "vega-embed": "^3.0.0-beta.0",
+    "vega-schema-url-parser": "^1.0.0-beta.2"
```
Comment (domoritz) [inline component/package.json]: I don't think you need to include the url parser.

**Comment by domoritz** [comment]
Can you try vega-embed beta 5?

**Comment by gnestor** [comment]
I installed beta5 and copied and pasted the dependencies from the beta5 source and `npm install`ed. That's when this error appeared. After downgrading to beta4 and copying and pasting the dependencies from beta4, the error went away.

**Comment by gnestor** [comment]
...which leads me to believe that this error was introduced by the upgrading of one vega-embed's dependencies:

```
"vega": "3.0.0-beta.25",
"vega-lite": "2.0.0-alpha.4",
"d3-selection": "^1.0.4",
```

**Comment by domoritz** [comment]
I'm installing jupyterlab_vega right now.

**Comment by domoritz** [comment]
Are there instructions on how to to do dev builds for the js part?

**Comment by gnestor** [comment]
You can run `bash build.sh` to build both extensions (takes some time) or `cd nbextension` or `cd labextension` and run `npm run build` inside those directories. If you are testing changes to the `component` directory (where vega-embed lives), then you have to do an `npm install` in the extension directory.

**Comment by domoritz** [comment]
Okay, I'm getting slightly different errors and I don't really understand how the build system works. 

screen shot 2017-03-10 at 14 55 39

In Vega-Lite, we are currently [adding vega and vl to the window and require the compiled files to avoid having to run babel again](https://github.com/vega/vega-lite/blob/master/site/static/main.ts#L2). This is certainly not ideal and I'm looking into ways to clean this up right now. 

Can you explain how vega and vega-lite get added to the build?

**Comment by gnestor** [comment]
Are you running jupyterlab? Or classic notebook? I was running classic notebook.

For jupyterlab_vega, we are using webpack to build both extensions. The extension runs `npm install ../component` to install the React component that depends on vega-embed, then it runs `npm install` to install its dependencies, then it runs webpack on its entry file (`index.js` for nbextension and `plugin.js` for labextension).

**Comment by domoritz** [comment]
I'm using jupyter lab. The problem is that I don't have a good handle on how the build system works. We use browserify and remove vega and vega-lite from our bundle (https://github.com/vega/vega-embed/blob/master/package.json#L59) so things are different when you use webpack. I suspect that there is a problem with es6 syntax. `event-selector.js` comes from vega-lite importing an es6 module from vega. es6 uses `import`, rather than `require` and that seems to cause problems here. I hope this helps.

**Comment by gnestor** [comment]
Ya, I think you're right. The way to resolve this with webpack is to add the modules that are written in ES6 to an `include` array in the babel-loader config. It's currently excluding everything in the node_modules directory except for `jupyterlab_vega_react` (which is the React component in `/component`). So, which modules need to be transpiled?

**Comment by blink1073** [comment]
The next release of the extension builder will have Webpack 2.0 support, which can load ES6 modules. cf https://github.com/jupyterlab/extension-builder/pull/51

**Comment by gnestor** [comment]
Sweet! I just upgraded mimerender-cookiecutter to use webpack 2 👍

**Comment by gnestor** [comment]
@domoritz beta6 is working for me. I still need to add all of the dependencies including `d3-selection` and `vega-schema-url-parser` which is unexpected, but it's working. Thanks!

**Comment by domoritz** [comment]
Vega-Lite is written in typescript and we compile to es5. We import [`vega-parser/src/parsers/event-selector`](https://github.com/vega/vega-lite/blob/9cc6c980314bbf1452bb3d6b64f251e3990716ff/src/compile/selection/transforms/zoom.ts#L1), which is es6.

**Comment by domoritz** [comment]
> @domoritz beta6 is working for me. I still need to add all of the dependencies including d3-selection and vega-schema-url-parser which is unexpected, but it's working. Thanks!

`vega-schema-url-parser` may be useful for you as you can sue it to parse the `$schema` url from vega or vega-lite specifications to figure out what type they are and which version they are targetting.

**Comment by domoritz** [comment]
@gnestor Btw, we have a slack channel at https://vega-js.slack.com that you are welcome to join and talk about issues more directly.

**Comment by gnestor** [comment]
Got it! Well, as @blink1073 said, this should no longer be an issue once jupyterlab/extension-builder#51 and https://github.com/jupyterlab/mimerender-cookiecutter/pull/56 are merged.

**Comment by domoritz** [comment]
Hopefully.

I was hoping to talk to you about some general design directions at some point.

**Comment by gnestor** [comment]
Ok, I'll join the Slack and we can start a convo.

EDIT: Looks like I need an invite!

**Pull request closed without merging**